File: packages/openterms-py/openterms/canonical.py
"""ORS v0.1 canonicalization.

Implements Section 4 of the Open Receipt Specification v0.1
(https://github.com/jstibal/ors-spec/blob/main/ORS-v0.1.md): RFC 8785 JSON
Canonicalization Scheme plus recursive null-stripping from objects.

Provenance note. BUILD_BRIEF Step 2 instructs porting canonicalization from a
legacy ``server/core/canonical.ts`` file. That file is not present in this
repository. This implementation is written directly against the ORS v0.1 spec
and matches the behavior of the reference verifier ``verify.py`` in
``jstibal/ors-spec`` so that receipts produced here pass third-party
verification by construction. The future TypeScript port should achieve
cross-language parity by passing the same test vectors at
``tests/vectors/ors-v0.1/canonicalization.json``, not by chasing the missing
legacy file.

Corner-case decisions (the spec is silent or ambiguous on each; behavior here
matches ``verify.py``):

  * Null stripping applies to objects only. Nulls inside arrays are preserved.
  * Empty containers (``{}`` and ``[]``) survive even after their last key was
    null-stripped; they are never pruned.
  * No Unicode normalization. NFC and NFD inputs produce different bytes.
  * Floats are REJECTED at canonicalize() time. Python ``repr`` and
    JavaScript ``Number.prototype.toString`` do not agree on every IEEE-754
    double, so silent pass-through risks cross-language divergence. Encode
    monetary amounts as integer cents, not as floats.
  * Non-BMP object keys are REJECTED. They would sort differently in JS
    (UTF-16 code units) vs Python (Unicode code points), producing
    divergent canonical bytes.
  * Integers beyond ``Number.MAX_SAFE_INTEGER`` (2**53 - 1) are REJECTED
    for the same reason — JS Number cannot represent them exactly. Encode
    such values as strings.
  * Key sort is by Python's default string ordering (Unicode code point).
"""

from __future__ import annotations

import hashlib
import json
from typing import Any

DOMAIN_SEPARATOR = b"ORSv0.1\x00"

PAYLOAD_KEYS_REQUIRED = (
    "workspace_id",
    "agent_id",
    "action_type",
    "terms_url",
    "terms_hash",
    "timestamp",
    "pricing_version",
)

PAYLOAD_KEYS_SIGNED_ENVELOPE = (
    "receipt_id",
    "amount_charged",
    "created_at",
)

PAYLOAD_KEYS_OPTIONAL = (
    "action_context",
    "ors_version",
    "issuer",
    "provider",
    "decision",
    "request_binding",
)


class CanonicalizationError(ValueError):
    """Raised when a value cannot be canonicalized identically across languages."""


# Beyond this magnitude, JavaScript's Number cannot represent the integer
# exactly, so the JS port emits a different string than Python would.
# Reject at canonicalization time to avoid silent cross-language divergence.
_MAX_SAFE_INTEGER = 9007199254740991  # 2**53 - 1


def _validate(obj: Any) -> None:
    if isinstance(obj, dict):
        for k, v in obj.items():
            if not isinstance(k, str):
                raise CanonicalizationError(
                    f"Object keys must be strings; got {type(k).__name__}"
                )
            # Non-BMP keys would sort differently in JS (UTF-16 code units)
            # vs Python (Unicode code points), producing divergent canonical
            # bytes. Reject explicitly.
            for ch in k:
                if ord(ch) > 0xFFFF:
                    raise CanonicalizationError(
                        "Object key contains a non-BMP (supplementary-plane) "
                        f"character; not supported by ORS v0.1 canonicalization: {k!r}"
                    )
            _validate(v)
        return
    if isinstance(obj, list):
        for v in obj:
            _validate(v)
        return
    if isinstance(obj, bool):
        return  # bool is a subclass of int; check before int
    if isinstance(obj, int):
        if abs(obj) > _MAX_SAFE_INTEGER:
            raise CanonicalizationError(
                f"Integer {obj} exceeds JavaScript Number.MAX_SAFE_INTEGER; "
                "encode as a string instead"
            )
        return
    if isinstance(obj, float):
        # Floats cannot round-trip identically between Python's repr() and
        # JS's Number.prototype.toString in all cases. NaN/Infinity also
        # fail allow_nan=False downstream, but reject up-front for clearer
        # error messages.
        raise CanonicalizationError(
            f"Float values are not allowed in ORS v0.1 canonical receipts (got {obj!r})"
        )
    # str, None, and other JSON-leaf values are fine.


def strip_nulls(obj: Any) -> Any:
    if isinstance(obj, dict):
        return {k: strip_nulls(v) for k, v in obj.items() if v is not None}
    if isinstance(obj, list):
        return [strip_nulls(v) for v in obj]
    return obj


def canonicalize(payload: dict) -> bytes:
    _validate(payload)
    cleaned = strip_nulls(payload)
    canonical_str = json.dumps(
        cleaned,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
        allow_nan=False,
    )
    return canonical_str.encode("utf-8")


def canonical_hash(payload: dict) -> str:
    return hashlib.sha256(canonicalize(payload)).hexdigest()


def signing_input(payload: dict) -> bytes:
    """Return the 40-byte Ed25519 message: domain separator + raw SHA-256."""
    digest = hashlib.sha256(canonicalize(payload)).digest()
    return DOMAIN_SEPARATOR + digest


def build_payload(receipt: dict) -> dict:
    """Extract the signed payload from a full receipt envelope.

    Excludes Section 3c signature metadata (``canonical_hash``, ``signature``,
    ``key_id``). Optional fields are included only if present and non-null.
    """
    payload: dict = {}
    for k in PAYLOAD_KEYS_REQUIRED:
        if k not in receipt:
            raise ValueError(f"Missing required payload field: {k}")
        payload[k] = receipt[k]
    for k in PAYLOAD_KEYS_SIGNED_ENVELOPE:
        if k not in receipt:
            raise ValueError(f"Missing required signed envelope field: {k}")
        payload[k] = receipt[k]
    for k in PAYLOAD_KEYS_OPTIONAL:
        if k in receipt and receipt[k] is not None:
            payload[k] = receipt[k]
    return payload

File: packages/openterms-ts/src/canonical.ts
// ORS v0.1 canonicalization — TypeScript port of packages/openterms-py/openterms/canonical.py.
//
// Behavior parity is enforced by the shared vectors at
// tests/vectors/ors-v0.1/canonicalization.json (the canonical-parity test reads
// the exact same file the Python suite reads).
//
// Cross-language sort note: this implementation sorts object keys by the
// default JavaScript string comparator, which compares UTF-16 code units.
// Python's json.dumps(sort_keys=True) sorts by Unicode code point. The two
// orderings agree for any key whose characters are entirely within the Basic
// Multilingual Plane (U+0000 - U+FFFF). Supplementary-plane keys would sort
// differently across languages and silently produce divergent canonical
// output, so canonicalize() rejects them explicitly (see assertBmpKey).
//
// Numeric handling is also strict: floats, NaN, Infinity, and integers
// outside Number.MAX_SAFE_INTEGER cannot round-trip identically between
// Python and JavaScript, so they are rejected at canonicalization time
// rather than silently emitted as divergent bytes.

import { sha256 } from "@noble/hashes/sha2";
import { bytesToHex } from "@noble/hashes/utils";

export const DOMAIN_SEPARATOR = new Uint8Array([
  0x4f, 0x52, 0x53, 0x76, 0x30, 0x2e, 0x31, 0x00,
]);
// ASCII "ORSv0.1" + 0x00 = 8 bytes. Sanity: matches b"ORSv0.1\x00" in canonical.py.

export const PAYLOAD_KEYS_REQUIRED = [
  "workspace_id",
  "agent_id",
  "action_type",
  "terms_url",
  "terms_hash",
  "timestamp",
  "pricing_version",
] as const;

export const PAYLOAD_KEYS_SIGNED_ENVELOPE = [
  "receipt_id",
  "amount_charged",
  "created_at",
] as const;

export const PAYLOAD_KEYS_OPTIONAL = [
  "action_context",
  "ors_version",
  "issuer",
  "provider",
  "decision",
  "request_binding",
] as const;

export const SIGNATURE_METADATA_KEYS = [
  "canonical_hash",
  "signature",
  "key_id",
] as const;

type JsonValue =
  | string
  | number
  | boolean
  | null
  | JsonValue[]
  | { [k: string]: JsonValue };

export class CanonicalizationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "CanonicalizationError";
  }
}

function assertBmpKey(key: string): void {
  for (let i = 0; i < key.length; i++) {
    const code = key.charCodeAt(i);
    // Surrogate pair => supplementary-plane code point (U+10000+).
    if (code >= 0xd800 && code <= 0xdbff) {
      throw new CanonicalizationError(
        `Object key contains a non-BMP (supplementary-plane) character; not supported by ORS v0.1 canonicalization: ${JSON.stringify(key)}`,
      );
    }
  }
}

export function stripNulls(value: unknown): unknown {
  if (value === null) return null;
  if (Array.isArray(value)) {
    return value.map((v) => stripNulls(v));
  }
  if (typeof value === "object") {
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      if (v === null || v === undefined) continue;
      out[k] = stripNulls(v);
    }
    return out;
  }
  return value;
}

// RFC 8785 JSON Canonicalization Scheme, applied to a payload that has already
// been null-stripped. Matches json.dumps(sort_keys=True, separators=(",", ":"),
// ensure_ascii=False, allow_nan=False) in canonical.py.
function canonicalSerialize(value: unknown): string {
  if (value === null) return "null";
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "number") {
    if (Number.isNaN(value)) {
      throw new CanonicalizationError("NaN is not a valid canonical value");
    }
    if (!Number.isFinite(value)) {
      throw new CanonicalizationError(
        "Infinity is not a valid canonical value",
      );
    }
    if (!Number.isInteger(value)) {
      // Floats cannot round-trip identically between Python's repr() and
      // JavaScript's Number.prototype.toString in all cases. Reject rather
      // than emit divergent bytes. Negative zero falls under this branch
      // (it is not Number.isInteger) and is also rejected.
      throw new CanonicalizationError(
        `Float values are not allowed in ORS v0.1 canonical receipts (got ${value})`,
      );
    }
    if (!Number.isSafeInteger(value)) {
      throw new CanonicalizationError(
        `Integer ${value} exceeds Number.MAX_SAFE_INTEGER; encode as a string instead`,
      );
    }
    return value.toString(10);
  }
  if (typeof value === "string") {
    // JSON.stringify produces RFC 8259-compliant escaping. JCS (RFC 8785)
    // additionally requires that non-ASCII characters appear literally, not
    // as \uXXXX escapes. JSON.stringify in Node escapes only the mandatory
    // control characters and quote/backslash, leaving the rest literal, which
    // matches Python's ensure_ascii=False. Verified by vectors 8 and 11.
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return "[" + value.map((v) => canonicalSerialize(v)).join(",") + "]";
  }
  if (typeof value === "object") {
    const obj = value as Record<string, unknown>;
    const keys = Object.keys(obj).sort();
    const parts: string[] = [];
    for (const k of keys) {
      assertBmpKey(k);
      parts.push(JSON.stringify(k) + ":" + canonicalSerialize(obj[k]));
    }
    return "{" + parts.join(",") + "}";
  }
  throw new Error(`Cannot canonicalize value of type ${typeof value}`);
}

export function canonicalize(payload: unknown): Uint8Array {
  const cleaned = stripNulls(payload);
  const str = canonicalSerialize(cleaned);
  return new TextEncoder().encode(str);
}

export function canonicalString(payload: unknown): string {
  return canonicalSerialize(stripNulls(payload));
}

export function canonicalHash(payload: unknown): string {
  return bytesToHex(sha256(canonicalize(payload)));
}

export function signingInput(payload: unknown): Uint8Array {
  const digest = sha256(canonicalize(payload));
  const out = new Uint8Array(DOMAIN_SEPARATOR.length + digest.length);
  out.set(DOMAIN_SEPARATOR, 0);
  out.set(digest, DOMAIN_SEPARATOR.length);
  return out;
}

export function buildPayload(
  receipt: Record<string, JsonValue>,
): Record<string, JsonValue> {
  const payload: Record<string, JsonValue> = {};
  for (const k of PAYLOAD_KEYS_REQUIRED) {
    if (!(k in receipt)) {
      throw new Error(`Missing required payload field: ${k}`);
    }
    payload[k] = receipt[k] as JsonValue;
  }
  for (const k of PAYLOAD_KEYS_SIGNED_ENVELOPE) {
    if (!(k in receipt)) {
      throw new Error(`Missing required signed envelope field: ${k}`);
    }
    payload[k] = receipt[k] as JsonValue;
  }
  for (const k of PAYLOAD_KEYS_OPTIONAL) {
    if (k in receipt && receipt[k] !== null && receipt[k] !== undefined) {
      payload[k] = receipt[k] as JsonValue;
    }
  }
  // v0.2 optional signed fields — pass through if present.
  for (const k of ["terms_type", "terms_service", "terms_version"] as const) {
    if (k in receipt && receipt[k] !== null && receipt[k] !== undefined) {
      payload[k] = receipt[k] as JsonValue;
    }
  }
  return payload;
}
